ubuntu/main.cc

```diff
@@ -0,0 +1,257 @@
+  glClearColor(236, 239, 241, 0);
+  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+  glFlush();
+  glfwSwapBuffers(window);
+  glfwMakeContextCurrent(nullptr);
+}
+
+static bool RunFlutterEngine(GLFWwindow *window,
+                             const std::string &flutter_app_directory,
+                             const std::string &main_path,
+                             const std::string &icu_data_path) {
+  FlutterRendererConfig config = {};
+  config.type = kOpenGL;
+  config.open_gl.struct_size = sizeof(config.open_gl);
+  config.open_gl.make_current = GLFWMakeContextCurrent;
+  config.open_gl.clear_current = GLFWClearContext;
+  config.open_gl.present = GLFWPresent;
+  config.open_gl.fbo_callback = GLFWGetActiveFbo;
+  std::string assets_path = flutter_app_directory + "/build/flutter_assets";
+  std::string full_main_path = flutter_app_directory + "/" + main_path;
+  std::string packages_path = flutter_app_directory + "/.packages";
+  FlutterProjectArgs args = {};
+  args.struct_size = sizeof(FlutterProjectArgs);
+  args.assets_path = assets_path.c_str();
+  args.main_path =
+      main_path.empty() ? main_path.c_str() : full_main_path.c_str();
+  // Packages path is ignored if main_path is empty, so this can be added in
+  // as-is.
+  args.packages_path = packages_path.c_str();
+  args.icu_data_path = icu_data_path.c_str();
+  FlutterEngine engine = nullptr;
+  auto result =
+      FlutterEngineRun(FLUTTER_ENGINE_VERSION, &config, &args, window, &engine);
+  if (result != kSuccess || engine == nullptr) {
+    return false;
+  }
+  GLFWClearCanvas(window);
+  glfwSetWindowUserPointer(window, engine);
+  GLFWwindowSizeCallback(window, kInitialWindowWidth, kInitialWindowHeight);
+  return true;
+}
+
+static void printUsage(char *app_name) {
+  std::cout << "Usage: " << app_name
+            << " --flutter_app_directory=<path_to_dart_app_root> " << std::endl;
+  std::cout << "\t\t--main_path=<path_to_main_dart_file>" << std::endl;
+  std::cout << "\t\t--icu_data_path=<path_to_icu_file>" << std::endl;
+  std::cout << std::endl;
+  std::cout << "Required Arguments:" << std::endl;
+  std::cout
+      << "\t--flutter_app_directory\t Path to your flutter app's main directory"
+      << std::endl;
+  std::cout
+      << "\t--icu_data_path\t\t Path to an icudtl.dat file. If you've built "
+         "the "
+      << std::endl;
+  std::cout << "\t\t\t\t flutter engine source this can be found in, for "
+            << std::endl;
+  std::cout << "\t\t\t\t example: " << std::endl;
+  std::cout
+      << "\t\t\t\t <flutter_engine_root>/src/out/host_debug_unopt/icudtl.dat"
+      << std::endl;
+  std::cout << "Optional Arguments:" << std::endl;
+  std::cout << "\t--main_path\t\t Path to the flutter app's main dart file "
+               "relative "
+            << std::endl;
+  std::cout << "\t\t\t\t to `--flutter_app_directory` (typically lib/main.dart)"
+            << std::endl;
+}
+
+int main(int argc, char **argv) {
```

Comment:
I think the best way to cut this into library/ and example/ is probably to have library/ have a couple of public functions:
- Something like CreateFlutterWindow: Takes all the arguments necessary to start the Flutter engine, along with the window dimensions and title. Creates the window, binds all the callbacks, and starts the engine. (Alternately the window could still be passed in, but I think having it create the window is more consistent with the fact that all the window callbacks will be managed internally.)
- A matching DestroyFlutterWindow: Takes the window, and does both the engine shutdown and the window teardown.

The the main file just passes the specific paths to the example app and runs the run loop.

This is somewhat artificial at the moment, but if we start bringing up plugins or adding packing of any kind before moving away from GLFW having the division already set up will make things clearer. And as mentioned in the overview, I'd like to keep the conceptual breakdown similar across platforms.

Reply:
Done (mostly)!